### DISCUSSION TITLE
UndefinedFunction

### DISCUSSION THREAD

**Discussion by AnjaliPare** [NationalSecurityAgency/ghidra#4835] [Q&A]
Hi,

Please could you explain to me what "UndefinedFunction" in ghidra implies. 
I am writing ghidra scripts to perform some analysis; however these functions are not being detected by my code. Please could you tell me how I may be able to perform analysis on the UndefinedFunctions?

Thank you!
-Anjali

**Comment by ghidracadabra** [comment]
If you see `UndefinedFunction_xxxxxxxx` and a grey background in the decompiler window it means that the bytes at the current address have been disassembled but have not been assigned to a function.  The decompiler works on functions, so in such cases it applies a heuristic to determine a function body (there's more information in the Ghidra help under "Undefined Functions").

It could be the case that the guessed function displayed in the decompiler does actually correspond to a function in the program.  However, it might not.  For example, it might just be the code for one case of a switch table.  It's probably best to examine these addresses and try to assign them to a function before doing any function-based analysis.  

In the code browser, you can enable the "Overview" margin display (via the down arrow next to the X in the Listing window).  In this display, instructions which are not in functions correspond to a different color than instructions which are.

**Comment by AnjaliPare** [comment]
Thanks! Is there a way I can get a list of these UndefinedFunctions so that I can remove them from my analysis?

**Comment by ghidra1** [comment]
You are better off the other way around.  If you iterate over all function symbols there will not be any `UndefinedFunctions`.    This is only used when decompiling code that does not reside within the body of a defined function - thus the name.  They really should be considered a function until one is created.  Usually they exist on code when analysis is incomplete (i.e., evidence of a function start was not found).

**Comment by ghidra1** [comment]
I am converting this ticket to a Discussion topic.
